kolibri_explore_plugin/assets/src/modules/topicsRoot/handlers.js

```diff
@@ -181,40 +181,53 @@ export function showFilteredChannels(store) {
   );
 }
 
-export function searchChannels(store, search, kind) {
-  store.commit('CORE_SET_PAGE_LOADING', true);
-  return ContentNodeSearchResource.fetchCollection({
+export function searchChannelsOnce(store, search, kind) {
+  const searchPromise = ContentNodeSearchResource.fetchCollection({
     getParams: {
       search,
       kind,
       max_results: SEARCH_MAX_RESULTS,
     },
-  }).then(searchResults => {
-    const { rootNodes } = store.state.topicsRoot;
-    const { channel_ids, results } = searchResults;
-    const nodes = results.map(n => ({
-      ...n,
-      channel: rootNodes.find(c => c.id === n.channel_id),
-    }));
-    const promises = channel_ids.map(id => ChannelResource.fetchModel({ id }));
-    Promise.all(promises).then(collection => {
-      const channels = collection
-        .map(c => ({
-          ...c,
-          thumbnail: getChannelIcon(c),
-          title: c.name,
-          order: channel_ids.indexOf(c.id),
-        }))
-        .sort((a, b) => a.order - b.order);
-      store.commit('topicsRoot/SET_SEARCH_RESULT', {
-        ...store.state.topicsRoot.searchResult,
-        [kind]: {
-          ...searchResults,
-          channels: channels,
-          results: nodes,
-        },
-      });
-      store.commit('CORE_SET_PAGE_LOADING', false);
+  });
+  store.commit('topicsRoot/SET_LAST_SEARCH_PROMISE', { kind, searchPromise });
+  return new Promise(resolve => {
+    searchPromise.then(searchResults => {
+      const { lastSearchPromises } = store.state.topicsRoot;
+      if (searchPromise == lastSearchPromises[kind]) {
+        store.commit('topicsRoot/SET_LAST_SEARCH_PROMISE', { kind, searchPromise: null });
+        setSearchResults(store, searchResults, kind);
+        resolve(searchResults);
+      }
```

Comment:
Maybe it could be a good idea to add here an `else` clause with the promise `reject`, to free the resources.

Reply:
@danigm after some investigation, we are not catching the promise so there is no need for a reject. Since we are not maintaining a reference to previous promises they will be garbage-collected. It's explained in [the article I've followed](https://codeburst.io/cancelling-async-operations-in-vue-js-3d0f3c2de598) and also [here](https://stackoverflow.com/questions/36734900/what-happens-if-you-dont-resolve-or-reject-a-promise#:~:text=A%20promise%20is%20just%20an,just%20a%20regular%20Javascript%20object.).

But I already added the reject. Please let me know if you want it or if I should revert the last change

Reply:
Looks good to me.